server/src/main/java/io/crate/expression/scalar/ConcatFunction.java

```diff
@@ -80,10 +82,43 @@ public static void register(Functions.Builder module) {
                 .argumentTypes(DataTypes.UNTYPED_OBJECT.getTypeSignature(),
                     DataTypes.UNTYPED_OBJECT.getTypeSignature())
                 .returnType(DataTypes.UNTYPED_OBJECT.getTypeSignature())
-                .features(Feature.DETERMINISTIC)
+                .features(Feature.DETERMINISTIC, Feature.NON_NULLABLE)
                 .build(),
             ObjectMergeFunction::new
         );
+
+
+        // Operator versions of concat, the default(string) version differs
+        // as it will return null if any of the arguments is null
+        module.add(
+            Signature.builder(OPERATOR_NAME, FunctionType.SCALAR)
+                .argumentTypes(DataTypes.STRING.getTypeSignature(),
+                    DataTypes.STRING.getTypeSignature())
+                .returnType(DataTypes.STRING.getTypeSignature())
+                .features(Feature.DETERMINISTIC, Feature.NULLABLE)
+                .build(),
+            (signature, boundSignature) -> new StringConcatFunction(signature, boundSignature, true)
+        );
+        module.add(
+            Signature.builder(OPERATOR_NAME, FunctionType.SCALAR)
+                .argumentTypes(TypeSignature.parse("array(E)"),
+                    TypeSignature.parse("array(E)"))
+                .returnType(TypeSignature.parse("array(E)"))
+                .features(Feature.DETERMINISTIC, Feature.NON_NULLABLE)
+                .typeVariableConstraints(typeVariable("E"))
+                .build(),
+            ArrayCatFunction::new
+        );
+        module.add(
+            Signature.builder(OPERATOR_NAME, FunctionType.SCALAR)
+                .argumentTypes(DataTypes.UNTYPED_OBJECT.getTypeSignature(),
+                    DataTypes.UNTYPED_OBJECT.getTypeSignature())
+                .returnType(DataTypes.UNTYPED_OBJECT.getTypeSignature())
+                .features(Feature.DETERMINISTIC, Feature.NON_NULLABLE)
```

Comment:
```suggestion
                .features(Feature.DETERMINISTIC)
```

I think the nullability is conditional here - if I see it right this will return `null` if both arguments are null

Same for `array_cat` which now also returns null if all arguments are null

Reply:
Hm right, thought null will match the string version but this isn't true when the data type is set.